File: proctoring/app/proctoring/rules.py
"""
Violation rules engine for proctoring policy enforcement
"""
from dataclasses import dataclass
from typing import Dict, Optional
from datetime import datetime, timedelta

from app.core.config import settings
from app.core.logging import logger
from app.models.exam import ViolationTypeEnum, ViolationSeverityEnum


@dataclass
class ViolationPolicy:
    """Policy for a violation type"""
    max_duration_seconds: Optional[float]  # None for instant violations
    severity: ViolationSeverityEnum
    auto_submit_on_critical: bool = False


class ViolationRulesEngine:
    """Enforces violation rules and tracks violations"""

    VIOLATION_POLICIES: Dict[ViolationTypeEnum, ViolationPolicy] = {
        ViolationTypeEnum.NO_FACE: ViolationPolicy(
            max_duration_seconds=settings.MAX_NO_FACE_SECONDS,
            severity=ViolationSeverityEnum.HIGH,
            auto_submit_on_critical=True,
        ),
        ViolationTypeEnum.MULTIPLE_FACES: ViolationPolicy(
            max_duration_seconds=None,  # Instant violation
            severity=ViolationSeverityEnum.CRITICAL,
            auto_submit_on_critical=True,
        ),
        ViolationTypeEnum.LOOKING_AWAY: ViolationPolicy(
            max_duration_seconds=settings.MAX_LOOKING_AWAY_SECONDS,
            severity=ViolationSeverityEnum.MEDIUM,
        ),
        ViolationTypeEnum.HEAD_TURN: ViolationPolicy(
            max_duration_seconds=3.0,  # 3 seconds tolerance
            severity=ViolationSeverityEnum.MEDIUM,
        ),
    }

    def __init__(self):
        """Initialize violation tracker"""
        # Track ongoing violations: {violation_type: start_datetime}
        self.ongoing_violations: Dict[ViolationTypeEnum, datetime] = {}
        self.violation_count: int = 0
        self.warning_count: int = 0

    def check_violation(
        self,
        violation_type: ViolationTypeEnum,
        condition: bool,
        current_time: datetime = None,
    ) -> Optional[Dict]:
        """
        Check if a violation should be triggered based on condition and duration.
        
        Args:
            violation_type: Type of violation
            condition: Current condition (True = violation condition present)
            current_time: Current datetime
            
        Returns:
            Dict with violation details if triggered, None otherwise
        """
        if current_time is None:
            current_time = datetime.utcnow()

        policy = self.VIOLATION_POLICIES[violation_type]

        if not condition:
            # Condition no longer present, clear from tracking
            if violation_type in self.ongoing_violations:
                del self.ongoing_violations[violation_type]
            return None

        # Condition is present
        if violation_type not in self.ongoing_violations:
            # First time we're seeing this violation
            self.ongoing_violations[violation_type] = current_time

            # If instant violation (no max_duration), trigger immediately
            if policy.max_duration_seconds is None:
                return self._create_violation(violation_type, policy, 0)

            return None

        # We've been tracking this violation
        start_time = self.ongoing_violations[violation_type]
        duration = (current_time - start_time).total_seconds()

        if duration >= policy.max_duration_seconds:
            # Duration threshold exceeded, trigger violation
            return self._create_violation(violation_type, policy, duration)

        return None

    def _create_violation(
        self,
        violation_type: ViolationTypeEnum,
        policy: ViolationPolicy,
        duration_seconds: float,
    ) -> Dict:
        """Create violation record"""
        self.violation_count += 1

        if policy.severity != ViolationSeverityEnum.CRITICAL:
            self.warning_count += 1

        violation = {
            "violation_type": violation_type,
            "severity": policy.severity,
            "duration_seconds": duration_seconds,
            "timestamp": datetime.utcnow(),
        }

        logger.info(
            f"Violation triggered: {violation_type}",
            extra={
                "severity": policy.severity,
                "duration": duration_seconds,
            },
        )

        return violation

    def should_auto_submit(self) -> bool:
        """
        Check if exam should be auto-submitted based on violations.
        
        Returns:
            True if auto-submit threshold exceeded
        """
        # Auto-submit if critical violation occurred
        if self.violation_count > 0:
            # Check if any recent violation was critical
            # (In practice, this would be tracked more carefully)
            pass

        # Auto-submit if warning count exceeded
        if self.warning_count >= settings.MAX_WARNINGS:
            logger.warning(
                f"Auto-submit triggered: {self.warning_count} warnings exceed threshold"
            )
            return True

        return False

    def get_violation_summary(self) -> Dict:
        """Get current violation summary"""
        return {
            "total_violations": self.violation_count,
            "warnings": self.warning_count,
            "ongoing_violations": list(self.ongoing_violations.keys()),
        }

    def reset(self):
        """Reset violation tracking for new session"""
        self.ongoing_violations.clear()
        self.violation_count = 0
        self.warning_count = 0

File: proctoring/src/proctoring/rules.js
/**
 * Violation rules engine for proctoring policy enforcement
 */
const settings = require('../config/config');
const logger = require('../config/logger');
const { ViolationTypeEnum, ViolationSeverityEnum } = require('../models/exam');

// Violation policies
const VIOLATION_POLICIES = {
    [ViolationTypeEnum.NO_FACE]: {
        maxDurationSeconds: settings.MAX_NO_FACE_SECONDS,
        severity: ViolationSeverityEnum.HIGH,
        autoSubmitOnCritical: true,
    },
    [ViolationTypeEnum.MULTIPLE_FACES]: {
        maxDurationSeconds: null, // Instant violation
        severity: ViolationSeverityEnum.CRITICAL,
        autoSubmitOnCritical: true,
    },
    [ViolationTypeEnum.LOOKING_AWAY]: {
        maxDurationSeconds: settings.MAX_LOOKING_AWAY_SECONDS,
        severity: ViolationSeverityEnum.MEDIUM,
        autoSubmitOnCritical: false,
    },
    [ViolationTypeEnum.HEAD_TURN]: {
        maxDurationSeconds: 3.0, // 3 seconds tolerance
        severity: ViolationSeverityEnum.MEDIUM,
        autoSubmitOnCritical: false,
    },
};

/**
 * Create a new rules engine instance
 */
function createRulesEngine() {
    return {
        ongoingViolations: new Map(),
        violationCount: 0,
        warningCount: 0,
    };
}

/**
 * Check if a violation should be triggered based on condition and duration
 */
function checkViolation(engine, violationType, condition, currentTime = null) {
    if (currentTime === null) {
        currentTime = new Date();
    }

    const policy = VIOLATION_POLICIES[violationType];

    if (!condition) {
        // Condition no longer present, clear from tracking
        if (engine.ongoingViolations.has(violationType)) {
            engine.ongoingViolations.delete(violationType);
        }
        return null;
    }

    // Condition is present
    if (!engine.ongoingViolations.has(violationType)) {
        // First time we're seeing this violation
        engine.ongoingViolations.set(violationType, currentTime);
        logger.info(`Started tracking violation: ${violationType}, threshold: ${policy.maxDurationSeconds}s`);

        // If instant violation (no maxDuration), trigger immediately
        if (policy.maxDurationSeconds === null) {
            return createViolation(engine, violationType, policy, 0);
        }

        return null;
    }

    // We've been tracking this violation
    const startTime = engine.ongoingViolations.get(violationType);
    const duration = (currentTime.getTime() - startTime.getTime()) / 1000;
    logger.info(`Violation ${violationType}: duration=${duration.toFixed(1)}s / threshold=${policy.maxDurationSeconds}s`);

    if (duration >= policy.maxDurationSeconds) {
        // Duration threshold exceeded, trigger violation
        engine.ongoingViolations.delete(violationType);
        logger.warning(`VIOLATION TRIGGERED: ${violationType} after ${duration.toFixed(1)}s`);
        return createViolation(engine, violationType, policy, duration);
    }

    return null;
}

/**
 * Create violation record
 */
function createViolation(engine, violationType, policy, durationSeconds) {
    engine.violationCount++;

    if (policy.severity !== ViolationSeverityEnum.CRITICAL) {
        engine.warningCount++;
    }

    const violation = {
        violation_type: violationType,
        severity: policy.severity,
        duration_seconds: durationSeconds,
        timestamp: new Date(),
    };

    logger.info(`Violation triggered: ${violationType}`, {
        severity: policy.severity,
        duration: durationSeconds,
    });

    return violation;
}

/**
 * Check if exam should be auto-submitted based on violations
 */
function shouldAutoSubmit(engine) {
    if (engine.warningCount >= settings.MAX_WARNINGS) {
        logger.warning(`Auto-submit triggered: ${engine.warningCount} warnings exceed threshold`);
        return true;
    }
    return false;
}

/**
 * Get current violation summary
 */
function getViolationSummary(engine) {
    return {
        total_violations: engine.violationCount,
        warnings: engine.warningCount,
        ongoing_violations: Array.from(engine.ongoingViolations.keys()),
    };
}

/**
 * Reset violation tracking
 */
function resetEngine(engine) {
    engine.ongoingViolations.clear();
    engine.violationCount = 0;
    engine.warningCount = 0;
}

module.exports = {
    VIOLATION_POLICIES,
    createRulesEngine,
    checkViolation,
    shouldAutoSubmit,
    getViolationSummary,
    resetEngine,
};
